File: problems/437/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if not root:
            return 0
        
        ans, q = 0, deque([(root, [])])
        
        while q:
            node, vals = q.popleft()
            nvals = vals + [0]
            
            for i in range(len(nvals)):
                nvals[i] += node.val
                
                if nvals[i] == sum:
                    ans += 1
            
            if node.left: q.append((node.left, nvals))
            if node.right: q.append((node.right, nvals))
        
        return ans

File: problems/437/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        long long int ans = 0;
        unordered_map<long long int, int> um;
        
        dfs(root, targetSum, um, 0, ans);
        
        return ans;
    }
    
    void dfs(TreeNode* root, int k, unordered_map<long long int, int>& um, long long int curSum, long long int& ans) {
        if (not root) {
            return;
        }
        
        curSum += root->val;
        ans += (curSum == k);
        ans += um[curSum - k];
        um[curSum]++;
        
        dfs(root->left, k, um, curSum, ans);
        dfs(root->right, k, um, curSum, ans);
        
        um[curSum]--;
    }
};